### ISSUE TITLE
Distributions returning NaN at support boundary

### ISSUE THREAD

**Issue by neerajprad** [probtorch/pytorch#145]
The following distributions (I only checked distributions whose support is positive or real) end up returning NaN when we pass `Inf` or -`Inf` to `log_prob`. 

```
Chi2 example 1/2 sample = inf
FisherSnedecor example 1/3 sample = inf
Gamma example 2/2 sample = inf
Gumbel example 1/2 sample = -inf
MultivariateNormal example 1/4 sample = inf
MultivariateNormal example 1/4 sample = -inf
```

I think it was part of our initial design doc to ensure that we did not raise error or return NaN for values in the support of the distribution, and tried to ensure that Bernoulli and Categorical returned valid values at the boundaries. 

Doing this for all distributions may be challenging due to the following:

 - We may need to specially check for `inf` values or alternatively clamp values to some max value that depends on tensor type, both of which can get expensive for large tensors.
 - For the first approach above, the gradient will be undefined still, even if it might have a well defined value in reality (0).

Is this something that we would like to support (even in a restricted mode like when some flag like `_validate_args` is enabled)? 

I came across this issue when running NUTS on a model that has a gamma RV. cc. @fritzo, @alicanb, @apaszke

**Comment by apaszke** [comment]
Do we really treat `inf` as something in the support? I wouldn't. It's not a real number and would never be a result of sampling

**Comment by neerajprad** [comment]
> Do we really treat inf as something in the support? I wouldn't. It's not a real number and would never be a result of sampling

That's true. Even though sampling from the same class will not produce it, there are many cases where we might want to score some value that is produced through an external routine (like the leapfrog integrator in HMC), and these can easily generate `inf`. Our constraints checks for support will pass with `inf`, but it will likely fail elsewhere. 

I think it is fine to not support `inf` in the general case (in fact desirable to prevent computational overhead and complexity) and return `NaN`s. We just need to ensure that we are sending NaNs back, or returning a consistent error that the upstream client code can catch. What would be bad is if we have multiple failure patterns, and the client is forced to wrap up a model inside a general `try catch`. I have noticed that we are only returning `ValueError` so we should be fine for now, but I'll report back if I see any issues.

**Comment by apaszke** [comment]
What's the density at inf though? Do you want to return a limit, which is basically 0? I'd say that if you have a procedure that can generate infs/NaNs then it should take care of handling the extreme cases, because they probably won't happen in most other use cases.

Not sure what the failure patterns and `try catch` relate to. Can you please elaborate?

**Comment by neerajprad** [comment]
tldr: Things are pretty good on the PyTorch front. I am also convinced that the solution is for the client code to handle this, and PyTorch seems to handle NaNs consistently (at least on the few distributions that I have encountered this). Most other libraries also return NaNs for the scenarios I mentioned, so I am closing this question.

>  I'd say that if you have a procedure that can generate infs/NaNs then it should take care of handling the extreme cases

Agreed, I am convinced that's the only sustainable solution for this.

> What's the density at inf though? Do you want to return a limit, which is basically 0?

That would be one option, i.e. to return `-inf` as log prob, but finding out the `inf` entries in a tensor and doing this for each entry is probably not worth it (and the gradient will still be undefined, whereas it should be 0 in the limit). 

> Not sure what the failure patterns and try catch relate to. Can you please elaborate?

I meant that operations on these distributions instances will likely be chained in a model, e.g. sampling from a half cauchy, running some operations, and piping this in as an std parameter to a Normal. If during the course of a model run, we encounter NaN's, the best course of option is to return them back. e.g. sampling from `Normal(0, NaN)` should just return NaN tensors, or else throw a consistent error. The client can decide what to do - e.g. in the case of HMC's step size adaptation, we are using this to reduce our step size. However, if certain operations on these distribution instances return NaN and others throw a `RuntimeError` or a `ValueError` then the client will be forced to wrap the model run inside a generic `try catch` block, which is likely to mask all other inference issues as well. 

From what I have found with HMC, things are pretty good on the PyTorch front. We seem to return NaNs back or throw ValueError if distribution arg check is enabled. Both of these seem like good design choices!

**Comment by apaszke** [comment]
Great, thanks for the explanation!
